### ISSUE TITLE
Some apps need persistence--we should provide an opinionated approach 

### ISSUE THREAD

**Issue by anglinb** [probot/friction#6]
**tl;dr** 👋 While most apps can get all the data they need from the GitHub API, some need to persist data. We should provide an API to do so.

---

Probot is a pretty generic project that doesn't limit what you create. Additionally, it doesn't provide a hosting mechanism so I imagine probot apps will have a lot of different implementations. However, data persistence is something that many might need for their apps to work. With this in mind, I propose adding a probot helper that supports an in-memory key-value store and a series of key-value adapters that allow for persistence through an app restart by leveraging existing datastores like MySQL, Redis, MongoDB, etc.

I think key-value stores are both generic enough that most applications can use them and powerful enough to meet most basic data needs. Additionally, almost every data store can be used a key value store so writing adapters for the most common databases would provide support for almost all hosting environments. 

### Proposed API

`async probot.set(String key, Object value)` - This method will set the key to the result of calling `JSON.stringify` on `value`. Passing in a value of `null` will delete the key.

`async probot.get(String key)` - This method will return the result of calling `JSON.parse` on the value found in the datastore or undefined.

### Questions

- Should we be serializing data on behalf of the user or allowing them to do it manually? (We could provide two different APIs one for Strings and one for Objects)
- Should we provide an iteration API [like leveldb provides](https://github.com/Level/levelup#createReadStream)? 
- Is a key-value store the most effective persistence we can provide while maintaining modularity? Would it be better to simply provide something like `probot-mongodb` that simply exposed a db object on the robot?

/cc @jasonetco @stargator @probot/maintainers

**Comment by JasonEtco** [comment]
Thanks for getting this discussion started @anglinb 🎉

I see where the Probot setters and getters can be handy, but I'll write my thoughts for a more persistent form of external storage because I think it's a little more difficult to unpack.

To responsibly create an API around an external database, we need to consider the following:
* Wrappers for multiple types of DBs
* Handling DB errors
* DB authentication

Nothing that we would add can't already be done without this functionality being a part of Probot, so whatever we come up with will be an abstraction over generic, typical DB configuration. Things like connections, authentication, etc.

So, I think we have two options:

1. We write a series of `probot-dbname` extensions and use a `robot.use()` type thing to tell the Probot App that its using a given database, and bind the relevant functionality to `robot.db`.

2. We encourage users to write their own database functionality/wrappers outside of their Probot App and use them as modules.

I'm happy to make a repo or (a PR) for a MongoDB wrapper, with an API that looks like:

```js
// probot-mongo.js
const mongoose = require('mongoose');
const mongoUri = `mongodb://${process.env.DB_HOST}`;

mongoose.connect(mongoUri, {
  user: process.env.DB_USER,
  pass: process.env.DB_PASS,
  useMongoClient: true,
});
```

```js
// index.js
const probotMongo = require('./probot-mongo');
const NameSchema = require('./NameSchema');

module.exports = async robot => {
  robot.db.use(probotMongo, [NameSchema]);
  const names = await robot.db.names.find();
};
```

**Comment by bkeepers** [comment]
Thanks for starting this discussion @anglinb & @JasonEtco. Lack of persistence was an intentional decision, but it's a good time to revisit it.

Everything I've built so far with Probot uses GitHub for persistence. This works great in places where there is a GitHub feature that naturally acts as persistence: labels, statuses, repository contents, etc. Avoiding persistence forces some clever designs (both the good and the bad kind of clever). It also has some beneficial side-effects. The operational overhead of running an app is greatly reduced. Everything about the project is stored in one place. If I don't like the direction a Probot app is going, I can fork it and run my own version without worrying about losing data. Someone can build a better plugin that uses the same data.

This doesn't work in places where GitHub doesn't have features that naturally map to an app's persistence needs, or when your app is integrating GitHub and some other service and needs to store data for that association. @wilhelmpklopp and I are working on an app like that now and we're going to have to store data for that mapping. (Aside: I wish I could just store arbitrary metadata on GitHub for an issue, pull request, comment, branch, repo, org, etc.)

That being said, I think my preference for the opinionated approach for now is: "avoid persistence unless absolutely necessary", and document some strategies for avoiding it and recommendations for situations where it is necessary in the [best practices](https://probot.github.io/docs/best-practices/).

I'm not staunchly opposed to persistence, but I think once we offer it, the operational overhead increases dramatically, and it introduces a lot of complexity that we'll have to deal with. It's not insurmountable, but I just want to make sure we absolutely have to go down that path.

Thoughts? Either way, I think we should keep collecting real world use cases of Probot apps that need persistence and keep exploring what persistence looks like for those apps.

**Comment by anglinb** [comment]
Thanks for following up @bkeepers! Wanted to quickly touch on a few points you made:
- I totally agree the opinion should be -> "avoid persistence unless absolutely necessary"
- +1 adding to best practices
- "operational overhead increases dramatically" I totally agree. By not providing something out of the box, we can try and push this off to individual developers. 

I've been googling around to see how other platforms deal with these challenges. Using slack as an example, I came across two frameworks (somewhat similar to probot though no where near as awesome 🤓):
- [Botkit](https://github.com/howdyai/botkit) - A bot framework for a few different platforms
- [Slapp](https://github.com/BeepBoopHQ/slapp) - A framework specifically for slack

These projects are both owned by companies that provide services for bots. They provide a some sort of AI thing(?) and hosting respectively. Because they have these offerings their platforms have different ways of dealing with persistence. 

Botkit provides a mongo storage engine: https://github.com/howdyai/botkit-storage-mongo. However, Slapp provides a key value store that is backed by their hosting platform. I think these two approaches make sense for different goals. Slapp locking you into your hosting platform is valuable. Botkit deferring complexity to the user makes sense b/c they don't want to distract from their platform. 

Looking at probot's current status, I agree with @bkeepers's take on dissuading people from storing content. The answer seems to be an awesome hosting platform for probot plugins 😉

**Comment by JasonEtco** [comment]
I 100% agree with the advantages of suggesting that users find a way out of requiring persisting data, but for when it is unavoidable we should have some clear suggestions around it. It could take the form of an external package that enables that functionality within Probot, or just some very well-written documentation with enough examples that its clear where to start from.

The nice thing about building a Probot App is that its really just a node module, and can be blended together with any other. We should make use of that to enable optional, modular bits of Probot functionality.

We should at the very least have a page on [`probot.github.io`](https://probot.github.io) kind of like this (I can start a PR if you'd like):

---

# Persistence

Generally speaking, adding database storage or persistence to your Probot App will greatly complicate your life. It's perfectly doable, but for most use-cases you'll be able to manage relevant data within GitHub issues, pull requests and your app's configuration file.

For when you absolutely do need external data storage, here are some examples using a few popular database solutions:

## MongoDB with Mongoose

## MySQL

## Redis

## Other things

**Comment by bkeepers** [comment]
👍 that all sounds great

**Comment by JasonEtco** [comment]
@bkeepers I'll get it going 🎉

**Comment by tcbyrd** [comment]
🚢 In https://github.com/probot/probot/pull/242

Docs hosted at https://probot.github.io/docs/persistence/
